## Assignment: LEGO Project

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.

## Assignment

In this assignment, you will create a number of functions to complete certain tasks to showcase Numpy/Pandas skills.


Let's start out by importing some standard imports.

In [2]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [3]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [5]:
#parts.info()
sets.describe

<bound method NDFrame.describe of            set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 c

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [1092]:
def data_count(df):
    return len(df)

In [1090]:
#data_count(sets)

18576

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [1071]:
Q2 = round(sets.num_parts.mean())
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [779]:
Q3 = int(sets.num_parts.median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [29]:
Q4 = sets.num_parts.max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [31]:
Q5 = sets[sets.num_parts == sets.num_parts.max()]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [39]:
def select_set_row(set_num):
    return sets[sets.set_num == set_num]

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [43]:
def select_set_numparts(num_pawts):
    return sets[sets.num_parts == num_pawts].sort_values(by = ['year', 'set_num'])

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [59]:
Q8A = sets.year.min()
Q8A

1949

In [60]:
Q8B = sets.year.max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [56]:
def select_set_year(year):
    return sets[sets.year == year].sort_values(by = 'set_num')

In [64]:
# rows that match minimum year
Q9A = select_set_year(Q8A)
Q9A

set_num                               name  year  theme_id  num_parts
11057  700.1-1         Extra-Large Gift Set (ABB)  1949       365        142
11070  700.2-1               Large Gift Set (ABB)  1949       365        178
11075  700.3-1              Medium Gift Set (ABB)  1949       365        142
11080  700.A-1              Small Brick Set (ABB)  1949       371         24
11081  700.B-1  Small Doors and Windows Set (ABB)  1949       371         12

In [63]:
# rows that match maximum year
Q9B = select_set_year(Q8B)
Q9B

set_num                                               name  \
336            10297-1                                     Boutique Hotel   
600            10781-1                          Spider-Man’s Techno Trike   
601            10782-1                      Hulk vs. Rhino Truck Showdown   
602            10783-1                        Spider-Man at Doc Ock’s Lab   
603            10784-1                     Spider-Man Webquarters Hangout   
...                ...                                                ...   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   
17207  9780794449247-1  DC Comics Super Heroes: Gotham City’s New Defe...   
17726    DOTS-SERIES-6                    DOTS Collectible Tiles Series 6   

       year  theme_id  num_parts  
336    2022       155       3068  
600    2022       706         59  
601    2022       706        110  
602    2022       706        131  
603    2022       706        155  
...     ...       ...        ...  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  
17207  2022       497          7  
17726  2022       688         16  

[183 rows x 5 columns]

**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [67]:
Q10 = int(sets.year.median())
Q10

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [77]:
Q11 = int(sets.groupby('year')[['set_num']].count().mean())
Q11

258

**Q12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [1048]:
# merge themes and sets
themes_sets = sets.merge(themes, left_on = 'theme_id', right_on = 'id', 
                        suffixes = [None, '_themes'])

themes_sets

set_num                                name  year  theme_id  \
0            001-1                               Gears  1965         1   
1            002-1      4.5V Samsonite Gears Motor Set  1965         1   
2           1030-1      TECHNIC I: Simple Machines Set  1985         1   
3           1038-1                  ERBIE the Robo-Car  1985         1   
4           1039-1                Manual Control Set 1  1986         1   
...            ...                                 ...   ...       ...   
18571  M20-2566-10  Modulex Box - 10x Window 1 x 6 x 6  1963       716   
18572  M20-2575-10  Modulex Box - 10x Window 1 x 7 x 5  1963       716   
18573  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
18574  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
18575  WEETABIX1-1                     Weetabix Castle  1970       414   

       num_parts   id name_themes  parent_id  
0             43    1     Technic        NaN  
1              3    1     Technic        NaN  
2            211    1     Technic        NaN  
3            120    1     Technic        NaN  
4             39    1     Technic        NaN  
...          ...  ...         ...        ...  
18571         10  716     Modulex        NaN  
18572         10  716     Modulex        NaN  
18573         10  716     Modulex        NaN  
18574         10  716     Modulex        NaN  
18575        471  414      Castle      411.0  

[18576 rows x 8 columns]

In [1049]:
def theme_by_year(year):
    return themes_sets[themes_sets.year == year][['id', 'name_themes']].sort_values('id').drop_duplicates(subset = ['id', 'name_themes']).reset_index(drop = True)

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [1043]:
#themes_sets[(themes_sets.name_themes == 'Supplemental') & (themes_sets.id == 473)]

set_num                                               name    year  \
5329   103-1                  4.5V Motor Set with Rubber Tracks  1969.0   
5330   107-1                                     4.5V Motor Set  1976.0   
5331  1129-1                                      Storage Cloth  1980.0   
5332  1715-1                                      Special Value  1995.0   
5333  1743-1                         Box of Standard Bricks, 5+  1995.0   
...      ...                                                ...     ...   
5423   940-1                             Flags, Signs and Trees  1973.0   
5424   941-1                             Black and Clear Bricks  1974.0   
5425   970-1                                    Lighting Bricks  1978.0   
5426   990-1  Trees and Signs (1971 version with granulated ...  1971.0   
5427   990-2  Trees and Signs (1969 version with old style t...  1969.0   

      theme_id  num_parts   id   name_themes  parent_id  
5329     473.0       24.0  473  Supplemental      465.0  
5330     473.0       31.0  473  Supplemental      465.0  
5331     473.0        1.0  473  Supplemental      465.0  
5332     473.0      400.0  473  Supplemental      465.0  
5333     473.0      400.0  473  Supplemental      465.0  
...        ...        ...  ...           ...        ...  
5423     473.0       19.0  473  Supplemental      465.0  
5424     473.0       48.0  473  Supplemental      465.0  
5425     473.0       17.0  473  Supplemental      465.0  
5426     473.0       12.0  473  Supplemental      465.0  
5427     473.0       11.0  473  Supplemental      465.0  

[99 rows x 8 columns]

In [1050]:
theme_by_year(1978)

id               name_themes
0     4            Expert Builder
1    67              Classic Town
2   189            Classic Castle
3   227                 Christmas
4   235                      4.5V
5   363                      Boat
6   364  Building Set with People
7   404                Hobby Sets
8   405                 Homemaker
9   412                   Airport
10  453                   Technic
11  469                 Basic Set
12  473              Supplemental
13  501                      Gear
14  505                 Basic Set
15  632                      Town

**Q13:** What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).

In [529]:
a = themes_sets.groupby('name_themes', as_index = False).count()
b = a.loc[a.set_num == a.set_num.max()]
Q13 = b.iloc[0, 0]
Q13

'Gear'

**Q14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [1051]:
# copy themes_sets
d2 = themes_sets.copy()

# Rename name column
d2 = d2.rename(columns = {'name':'name_sets'})

# subset of themes_sets
d2 = d2[['set_num', 'name_sets', 'year', 'num_parts', 'theme_id', 'name_themes']]

# create theme_by_name_function

def theme_by_name(theme_name):
    return d2[d2.name_themes == theme_name].sort_values(by = 'set_num').reset_index(drop = True)

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q13`.  You should return 1,904 sets that match that theme name.

In [1052]:
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [532]:
def theme_by_setnum(set_num):
    return d2.loc[d2.set_num == set_num, 'name_themes'].item()

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [533]:
theme_by_setnum('11138-1')

'Gear'

**Q16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q16`.  Instead you will use this information for later questions.

In [534]:
mf = themes_sets.loc[themes_sets.name == 'Millennium Falcon']
mf1 = mf[mf.num_parts == mf.num_parts.max()]
mf1

set_num               name  year  theme_id  num_parts   id  \
12884  75192-1  Millennium Falcon  2017       171       7541  171   

                     name_themes  parent_id  
12884  Ultimate Collector Series      158.0

**Q17:** 
- How many minifigs are contained in the set from `Q16`?  Save this as `Q17` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [379]:
# tables to join: sets, inventories, inventory_minifigs, minifigs ???
# join sets - inventories - inventory_minifigs -mf1

merged = pd.merge(pd.merge(sets, inventories, on = 'set_num'), inventory_minifigs, left_on = 'id', right_on = 'inventory_id')

# merge above on mf1 on set_num and/or name 
m2 = merged.merge(mf1, on = 'set_num')
m2.merge(minifigs, on = 'fig_num')

Q17 = len(m2)
Q17

8

**Q18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [660]:
#mini2[mini2.set_num == '10217-1']
merged

set_num                             name  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts     id  version  inventory_id     fig_num  quantity  
0             12   5087        1          5087  fig-000225         1  
1             12   5087        1          5087  fig-000834         1  
2             12   5087        1          5087  fig-004401         1  
3             12   1414        1          1414  fig-000020         2  
4             12   4609        1          4609  fig-001127         2  
...          ...    ...      ...           ...         ...       ...  
18139        147  13813        1         13813  fig-001361         1  
18140          6  34919        1         34919  fig-010738         1  
18141        413  36159        1         36159  fig-004483         1  
18142        413  36159        1         36159  fig-004484         1  
18143        413  36159        1         36159  fig-010456         1  

[18144 rows x 10 columns]

In [515]:
# merge inventories, inventory_minifigs, minifigs
mini = pd.merge(pd.merge(inventories, inventory_minifigs, left_on = 'id', right_on = 'inventory_id'), 
         minifigs, left_on = 'fig_num', right_on = 'fig_num')

# rename 'name' to 'fig_name'
mini = mini.rename(columns = {'name': 'fig_name', 'num_parts': 'fig_num_parts'})

# subset required columns
mini2 = mini[['set_num', 'fig_num', 'fig_name', 'fig_num_parts', 'version']]

# create function 
def get_minifigs(set_nums, vers):
    return mini2[(mini2.version == vers) & (mini2.set_num == set_nums)].drop(['version'], axis = 1).sort_values('fig_num').reset_index(drop = True)

**Q19:** Using the above function, output information about the minifigs from the set discussed in `Q16`.  Save this output as `Q19`.

In [516]:
Q19 = get_minifigs('75192-1', 1)
Q19

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [512]:
get_minifigs('75192-1', 1)

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Q20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [790]:
## use themes_sets -- themes/sets
minifig_theme = themes_sets.merge(mini, on = 'set_num')

minifig_theme2 = minifig_theme[['fig_num', 'fig_name', 'theme_id']]

# create function
def minifigs_from_themes(theme_id):
    return minifig_theme2[minifig_theme2.theme_id == theme_id].drop_duplicates('fig_num').sort_values(by = 'fig_num').drop('theme_id', axis = 1).reset_index(drop = True)

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q16`.  There should be 149 minifigs output for that respective theme.

In [791]:
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [656]:
#minifig_theme[minifig_theme.fig_num == 'fig-001039']
inventory_parts[inventory_parts.inventory_id == 1]

inventory_id        part_num  color_id  quantity is_spare
0             1        48379c01        72         1        f
1             1           48395         7         1        f
2             1  stickerupn0077      9999         1        f
3             1         upn0342         0         1        f
4             1         upn0350        25         1        f

In [657]:
colors[colors.id == 1]

id  name     rgb is_trans
2   1  Blue  0055BF        f

In [815]:
# Subset required colums from previosu variable
fig_sets = minifig_theme[['set_num', 'version', 'name', 'fig_num', 'fig_name']]

# rename 'name' columns
fig_sets = fig_sets.rename(columns = {'name':'name_sets'})

# create function
def sets_from_minifig(minifig_num):
    return fig_sets[fig_sets.fig_num == minifig_num].sort_values(by = ['set_num', 'version']).reset_index(drop = True)

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [814]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q22:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [1085]:
# join iv parts on parts then join on colors -- two left join 
pawts = pd.merge(inventory_parts, parts)

pawts = pawts.rename(columns = {'name':'name_part'})

p2 = pd.merge(pawts, colors, how = 'outer', left_on = 'color_id', right_on = 'id')

# merge sets with inventories to get version
d3 = d2.merge(inventories, on = 'set_num')
d3 = d3[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'id']]

# merge d3 and p2 somehow * insert smiley face *
full_merge = pd.merge(d3, p2, how = 'outer', left_on = 'id', right_on = 'inventory_id')

full_merge = full_merge.rename(columns = {'name':'name_color'})

# subset columsn 
full_merge = full_merge[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 
           'name_part', 'name_color', 'quantity', 'is_spare']]

full_merge['year'] = full_merge['year'].fillna(0).astype(int)
full_merge['version'] = full_merge['version'].fillna(0).astype(int)
full_merge['quantity'] = full_merge['quantity'].fillna(0).astype(int)

def set_parts(set_num, vers):
    return full_merge[(full_merge.version == vers) & (full_merge.set_num == set_num)].sort_values(by = ['part_num', 'name_color']).reset_index(drop = True)

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q16`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [1086]:
set_parts('75192-1', 1)

set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray         3        f  
1    Light Bluish Gray        12        f  
2     Dark Bluish Gray         5        f  
3                White        22        f  
4                  Tan         3        f  
..                 ...       ...      ...  
725  Light Bluish Gray        70        f  
726              Black         6        f  
727   Dark Bluish Gray        25        f  
728  Light Bluish Gray         2        f  
729  Light Bluish Gray         3        f  

[730 rows x 10 columns]

  **Q23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q23`. 

In [806]:
Q23 = colors.name.sort_values().unique().tolist()
Q23

['Aqua',
 'Black',
 'Blue',
 'Blue-Violet',
 'Bright Green',
 'Bright Light Blue',
 'Bright Light Orange',
 'Bright Light Yellow',
 'Bright Pink',
 'Brown',
 'Chrome Antique Brass',
 'Chrome Black',
 'Chrome Blue',
 'Chrome Gold',
 'Chrome Green',
 'Chrome Pink',
 'Chrome Silver',
 'Copper',
 'Coral',
 'Dark Azure',
 'Dark Blue',
 'Dark Blue-Violet',
 'Dark Bluish Gray',
 'Dark Brown',
 'Dark Flesh',
 'Dark Gray',
 'Dark Green',
 'Dark Orange',
 'Dark Pink',
 'Dark Purple',
 'Dark Red',
 'Dark Tan',
 'Dark Turquoise',
 'Earth Orange',
 'Fabuland Brown',
 'Fabuland Orange',
 'Flat Dark Gold',
 'Flat Silver',
 'Flesh',
 'Glitter Trans-Clear',
 'Glitter Trans-Dark Pink',
 'Glitter Trans-Light Blue',
 'Glitter Trans-Neon Green',
 'Glitter Trans-Orange',
 'Glitter Trans-Purple',
 'Glow In Dark Opaque',
 'Glow In Dark Trans',
 'Glow in Dark White',
 'Green',
 'Lavender',
 'Light Aqua',
 'Light Blue',
 'Light Bluish Gray',
 'Light Flesh',
 'Light Gray',
 'Light Green',
 'Light Lime',
 'Light 

**Q24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q24` (as an integer).

In [746]:
Q24 = colors.loc[colors.is_trans == 't']['is_trans'].count()
Q24

38

**Q25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [1057]:
# merge colors and elemets tables 
p3 = pd.merge(colors, elements, how = 'left', left_on = 'id', right_on = 'color_id')

# create function 
def get_part_colors(part_nums):
    return p3[p3.part_num == part_nums].sort_values('color_id').drop_duplicates('name').reset_index(drop = True)[['id', 'name', 'rgb', 'is_trans']]
